# 章节9(课时79~课时)    
## 异常概念   
> **什么是异常？**   
> &ensp;&ensp;&ensp;&ensp;所谓异常是指在程序运行的过程中发生的一些不正常事件。（如除0溢出，数组下标越界，所要读取的文件不存在）。    
> **导致的后果？**    
> &ensp;&ensp;&ensp;&ensp;Java程序的执行过程中如出现异常事件，可以生成一个异常类对象，该异常对象封装了异常事件的信息，并将其被提交给java运行时系统，这个过程称为抛出异常，不处理的话会直接导致程序直接中断。   
> **如何防止程序中断**    
> &ensp;&ensp;&ensp;&ensp;设计良好的程序应该在程序异常发生时提供处理这些异常的方法，使得程序不会因为异常的发生而阻断或产生不可预见的结果。   
## 异常分类   
<img src="http://img.my.csdn.net/uploads/201211/27/1354020417_5176.jpg">  

|异常类型|包括|来源|处理|
|:-:|:-:|:-:|:-:|
|受查异常<br/>checkedException|Exception及其子类(不包括<br/>RuntimeException及其子类)|由代码控制能力之<br/>外的因素导致的运<br/>行时错误|必须要理解，否则<br/>通不过编译|
|非受查异常<br/>uncheckedException|Error和RuntimeException<br/>及其子类|RuntimeException<br/>一般代表编程错误|可以不用处理|
## Java异常处理机制    
> **Java的异常是通过两种机制来处理的**    
> &ensp;&ensp;&ensp;&ensp;**捕获：**`try-catch-finally`    
> &ensp;&ensp;&ensp;&ensp;**抛出：**`throw,throws`     
## try-catch-finally    
> **try{}语句块中放的是要检测的java代码，可能有会抛出异常，也可能会正常执行。**    
> **catch(异常类型){}块是当Java运行时系统接收到try块中所抛出异常对象时，会寻找能处理这一异常catch块来进行处理（可以有多个catch块）。**    
> **finally{}不管系统有没有抛出异常都会去执行，一般用来释放资源。除了在之前执行了System.exit(0)**    
```
import java.util.Scanner;
import java.util.InputMismatchException;
public class ExceptionDemo1{
    public static void main(String []args){
        System.out.println("请输入一个数字");
        Scanner input=new Scanner(System.in);
        int res=0;
        try{
            //return;
            System.exit(0);
            int number=input.nextInt();
            res=10/number;
        }catch(InputMismatchException e){
            System.out.println(e.getMessage());
            e.printStackTrace();
        }catch(ArithmeticException e){
            System.out.println(e.getMessage());
            e.printStackTrace();
        }catch(Exception e){
            System.out.println(e.getMessage());
        }finally{
            //释放资源，比如关闭打开的文件，删除一些临时文件等
            System.out.println("结果为："+res);
        }
    }
}
```
